### ISSUE TITLE
Improve docs and tests for model.save() / model.load()

### ISSUE THREAD

**Issue by dniku** [hill-a/stable-baselines#301]
- [ ] Document which state is preserved and not preserved when model is reloaded
- [ ] Add tests to ensure that the state that is documented to be preserved is actually preserved

---

Original question:

Is there a test to make sure that if I stop training, save the model, reload it from file, and continue training, the result will be identical compared to what would happen if I didn't stop training at all? I am mostly concerned about obscure state such as optimizer's internal variables.

**Comment by araffin** [comment]
Hello,

> I am mostly concerned about obscure state such as optimizer's internal variables.

that is a good question. I think there are different things that are not currently stored which prevent a perfect recover from previous training:
- the content of replay buffer is not saved for off-policy methods (for disk usage reason)
- the state of the optimizer (e.g. Adam) is not saved
- the learning rate schedule may be reset when reloading a model / resuming training 

However, despite those caveats, I did not experience huge drop in performance (yet) when resuming training.

**Comment by Pastafarianist** [comment]
Hm. Does the following sound good:

1. Close this issue
2. Open more issues:
    1. "Document which state is preserved and not preserved when model is reloaded"
    2. "Add tests to ensure that the state that is documented to be preserved is actually preserved"

?

**Comment by araffin** [comment]
I don't know, I would maybe keep that issue, what do you think @hill-a , @erniejunior or @AdamGleave ?

**Comment by AdamGleave** [comment]
I think it makes sense to keep this issue open but maybe break it into a [task list](https://help.github.com/en/articles/about-task-lists).

**Comment by YeeCY** [comment]
Any update for resuming training in an unexpected crash case?
And I also found that set `reset_num_timesteps=False` in the `learn` function cannot create a continuous Tensorboard curve after broken, any suggestion?

This is my code

    env = gym.make('CartPole-v1')

    trpo = TRPO(
        MlpPolicy, env,
        timesteps_per_batch=512,
        max_kl=0.001,
        cg_iters=10,
        cg_damping=1e-3,
        entcoeff=0.0,
        gamma=0.99,
        lam=1.0,
        vf_stepsize=1e-4,
        vf_iters=3,
        tensorboard_log='./tmp_tb_logs',
        verbose=1, n_cpu_tf_sess=None)
    trpo.learn(total_timesteps=5e4)
    trpo.save('./tmp_models/trpo_cartpole')

    del trpo

    trpo = TRPO.load('./tmp_models/trpo_cartpole', env=env)
    trpo.learn(total_timesteps=10e4, reset_num_timesteps=False)

**Comment by jtromans** [comment]
+1 to continuation of tensorboard logging. when reset_num_timesteps=False

**Comment by mpgussert** [comment]
Has there been any update to this? Checkpoint files are still missing significant data required to resume training, defeating the purpose of a checkpoint file.

**Comment by Miffyli** [comment]
@mpgussert 

Not in stable-baselines, but in [stable-baselines3](https://github.com/DLR-RM/stable-baselines3) the code saves all parameters (including optimizer parameters and replay buffer, if so desired).

**Comment by mpgussert** [comment]
yep!  i realized very quickly that i was mistaken.  sorry!

**Comment by araffin** [comment]
Fixed in SB3: https://github.com/DLR-RM/stable-baselines3/issues/435
